File: 백준/Gold/13460. 구슬 탈출 2/구슬 탈출 2.py
import sys
input = sys.stdin.readline
from collections import deque

m, n = map(int, input().split())
board = [list(input().strip()) for _ in range(m)]

for i in range(m):
    for j in range(n):
        if board[i][j] == 'R':
            r1,r2 = i,j
        elif board[i][j] == 'B':
            b1,b2 = i,j

dy = [-1,1,0,0]
dx = [0,0,-1,1]


q = deque()
q.append((r1,r2,b1,b2,0))
visited = []
visited.append((r1,r2,b1,b2))

ans = False

def move(x, y, dx, dy):
    cnt = 0
    while board[x+dx][y+dy] != '#' and board[x][y] != 'O':
        x += dx
        y += dy
        cnt += 1
    return x, y, cnt

while q:
    rx, ry, bx, by, cnt = q.popleft()

    # cnt가 10 초과일 경우 -1 호출
    if cnt>10:
        break

    # R이 O에 다다를 경우 cnt를 호출
    if board[rx][ry] == 'O':
        print(cnt)
        ans = True
        break

    for i in range(4):
        nrx, nry, r_move = move(rx, ry, dx[i], dy[i])
        nbx, nby, b_move = move(bx, by, dx[i],dy[i])
        # O에 구슬이 도달하는 경우는
        # 1. 빨간색만 도달한다.
        # 2. 파란색만 도달한다.
        # 3. 빨간색, 파란색이 동시에 도달한다.
        # 이 때 1번을 제외하고는 정답이 아니기 때문에 파란색이 O에
        # 도달 시 queue에 넣어주지 않기 위함임.
        if board[nbx][nby] == 'O':
            continue

        # R, B가 같은 좌표에 도달 시 더 많이 움직인 공이 뒤에 있었던 것
        # 더 많이 움직인 공을 한 좌표 뒤로 움직여준다.
        if nrx == nbx and nry == nby:
            if r_move>b_move:
                nrx -=dx[i]
                nry -=dy[i]
            else:
                nbx -=dx[i]
                nby -=dy[i]

        # R,B가 같은 좌표로 다시 돌아오게 되면 경우가 많아지므로 가지치기
        if not (nrx,nry,nbx,nby) in visited:
            visited.append((nrx,nry,nbx,nby))
            # queue에 새로운 좌표를 입력하여 bfs 진행
            q.append((nrx,nry,nbx,nby,cnt+1))

if ans == False:
    print(-1)

File: 백준/Gold/13460. 구슬 탈출 2/구슬 탈출 2.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Set;
import java.util.StringTokenizer;

public class Main {
	static int n;
	static int m;
	static char[][] arr;
	static int[] dx = { -1, 1, 0, 0 };
	static int[] dy = { 0, 0, -1, 1 };

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());

		int rx1 = 0;
		int ry1 = 0;
		int bx1 = 0;
		int by1 = 0;

		arr = new char[n][m];
		for (int i = 0; i < n; i++) {
			String word = br.readLine();
			for (int j = 0; j < m; j++) {
				char c = word.charAt(j);
				arr[i][j] = c;

				if (c == 'R') {
					rx1 = i;
					ry1 = j;
				}
				if (c == 'B') {
					bx1 = i;
					by1 = j;
				}
			}
		}

		Queue<int[]> q = new LinkedList<>();
		Set<Integer> visited = new HashSet<>();

		q.offer(new int[] { rx1, ry1, bx1, by1, 0 });
		visited.add(hash(rx1, ry1, bx1, by1));

		int rx, ry, bx, by, cnt;
		int ans = -1;

		while (!q.isEmpty()) {
			int[] qPoll = q.poll();
			rx = qPoll[0];
			ry = qPoll[1];
			bx = qPoll[2];
			by = qPoll[3];
			cnt = qPoll[4];

			if (arr[rx][ry] == 'O') {
				ans = cnt;
				break;
			}

			if (cnt > 10) {
				ans = -1;
				break;
			}

			for (int i = 0; i < dx.length; i++) {
				int[] rPoll = move(rx, ry, dx[i], dy[i]);
				int nrx = rPoll[0];
				int nry = rPoll[1];
				int rmove = rPoll[2];

				int[] bPoll = move(bx, by, dx[i], dy[i]);
				int nbx = bPoll[0];
				int nby = bPoll[1];
				int bmove = bPoll[2];

				if (arr[nbx][nby] == 'O') {
					continue;
				}

				if (nrx == nbx & nry == nby) {
					if (rmove > bmove) {
						nrx -= dx[i];
						nry -= dy[i];
					} else {
						nbx -= dx[i];
						nby -= dy[i];
					}
				}

				boolean isExist = visited.contains(hash(nrx, nry, nbx, nby));

				if (!isExist) {
					q.offer(new int[] { nrx, nry, nbx, nby, cnt + 1 });
					visited.add(hash(nrx, nry, nbx, nby));
				}
			}
		}
		sb.append(ans);
		System.out.println(sb.toString());
	}

	static int hash(int... values) {
		int result = 4;
		for (int value : values) {
			result = 8 * result + value;
		}
		return result;
	}

	static int[] move(int x, int y, int dx, int dy) {
		int cnt = 0;
		while (arr[x + dx][y + dy] != '#' & arr[x][y] != 'O') {
			x += dx;
			y += dy;
			cnt += 1;
		}
		return new int[] { x, y, cnt };
	}
}